### PR TITLE
FIR IDE: AddToString quickfix

### PR DESCRIPTION
**Pull request by tgeng**

### CODE REVIEW

File: idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddToStringFixFactories.kt [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+object AddToStringFixFactories {
+    @OptIn(ExperimentalStdlibApi::class)
+    private fun KtAnalysisSession.getFixes(element: PsiElement?, expectedType: KtType, actualType: KtType): List<AddToStringFix> {
+        if (element !is KtExpression) return emptyList()
+        return buildList {
+            if (expectedType.isString || expectedType.isCharSequence) {
+                add(AddToStringFix(element, false))
+                if (expectedType.isMarkedNullable && actualType.isMarkedNullable) {
+                    add(AddToStringFix(element, true))
+                }
```
Comment (jsjeon) [inline idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddToStringFixFactories.kt:25]: So, if expected type and actual type are nullable, a user will get two fixes, one with `$0.toString()` and the other with `$0?.toString()`? Should this be:
```kt
add(AddToStringFix(element, expectedType.isMarkedNullable && actualType.isMarkedNullable)
```
?

File: idea/idea-frontend-independent/src/org/jetbrains/kotlin/idea/quickfix/AddToStringFix.kt [hunk 2]
```diff
@@ -8,28 +8,29 @@ package org.jetbrains.kotlin.idea.quickfix
...
 import org.jetbrains.kotlin.psi.KtExpression
-import org.jetbrains.kotlin.psi.KtFile
 import org.jetbrains.kotlin.psi.KtPsiFactory
 import org.jetbrains.kotlin.psi.createExpressionByPattern
 import org.jetbrains.kotlin.psi.psiUtil.endOffset
 
-class AddToStringFix(element: KtExpression, private val nullable: Boolean) : KotlinQuickFixAction<KtExpression>(element),
-    LowPriorityAction {
+class AddToStringFix(element: KtExpression, private val useSafeCallOperator: Boolean) :
+    KotlinCrossLanguageQuickFixAction<KtExpression>(element), KotlinUniversalQuickFix, LowPriorityAction {
```
Comment (punzki) [inline idea/idea-frontend-independent/src/org/jetbrains/kotlin/idea/quickfix/AddToStringFix.kt:21]: What is the purpose of changing the superclasses here? Usually we just change to `KotlinPsiOnlyQuickFixAction`

Comment (tgeng) [reply idea/idea-frontend-independent/src/org/jetbrains/kotlin/idea/quickfix/AddToStringFix.kt:21]: Using `KotlinPsiOnlyQuickFixAction` sounds good. Whatever is used here doesn't seem to matter.

File: idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddToStringFixFactories.kt (hunk 1 above)
Comment (tgeng) [reply idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddToStringFixFactories.kt:25]: This is simply copying logic from FE1.0 at https://github.com/JetBrains/kotlin/blob/85cbea70bfc7b5814305210b8cf6286bfc6bda2a/idea/src/org/jetbrains/kotlin/idea/quickfix/QuickFixFactoryForTypeMismatchError.kt#L125.

Comment (punzki) [reply idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddToStringFixFactories.kt:25]: This makes sense to me, if you have code like:
```kt
fun foo(s: String?) {}
fun test(i: Int?) {
  foo(i)
}
```
The call can be changed to either `foo(i.toString())` (extension function that returns `"null"`) or `foo(i?.toString())` (simply returns `null`)

Comment (punzki) [reply idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddToStringFixFactories.kt:25]: Oops, comment streams crossed. 🙂  But you know, it _could_ be confusing to the user to offer `i.toString()`. Most of the time users may not want the `"null"` return value there (unless you are trying to render a string), and I don't think it's common knowledge that there is an `Any?.toString()` extension that return `"null"`. So I can see an argument for what @jsjeon is suggesting.

Comment (jsjeon) [reply idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddToStringFixFactories.kt:25]: Thanks for the example, @punzki . Didn't think about this that way, huh. I'd blame extension functions (`toString`) that allow nullable receiver and silently returns `null` back again. :p

**Review by jsjeon — APPROVED** [review]

**Review by darthorimar — APPROVED** [review]

**Comment by darthorimar** [comment]
Merged manually, thanks!

**Pull request closed without merging**